Rewrite

RealOrangeOne opened the pull request:
- [x] Remove old code
- [x] Build Haste Map
- [x] Add mocks for native modules
- [x] Add component mocks
- [x] Add extra require definitions
- [x] Add unit tests for mocking
- [x] Increase coverage
- [x] Basic tests for react-native core components
- [ ] Basic tests for react-native core APIs

Lasted Checked Version: 0.42

RealOrangeOne committed:
Remove old source

RealOrangeOne committed:
Update dependancies

RealOrangeOne committed:
Add babel build

RealOrangeOne committed:
Add haste map generator

RealOrangeOne committed:
Build file on post install

RealOrangeOne committed:
Use require, dummy!

RealOrangeOne committed:
Remove old test files

RealOrangeOne committed:
Add prepublish

RealOrangeOne committed:
dont commit the map fool!

RealOrangeOne committed:
Add mock file

RealOrangeOne committed:
use imports

RealOrangeOne committed:
use built script

RealOrangeOne committed:
Start on mocking code

RealOrangeOne committed:
Make haste script node 4 compatable

RealOrangeOne committed:
wrong lib...

RealOrangeOne committed:
Disable var check on haste builder

RealOrangeOne committed:
Add extra dependandies for node 4

RealOrangeOne committed:
Add remaining mock library

RealOrangeOne committed:
add some basic tests

RealOrangeOne committed:
Update lint config

RealOrangeOne committed:
add tests for mock component script

RealOrangeOne committed:
Fix linter

RealOrangeOne committed:
Better tests

RealOrangeOne committed:
fix typo

RealOrangeOne committed:
remove pointless test

RealOrangeOne committed:
Add coverage

RealOrangeOne committed:
Added nycrc file, woops

RealOrangeOne committed:
More tests on global

RealOrangeOne committed:
Add better tests to haste map

RealOrangeOne committed:
Add longer timeout for tests

RealOrangeOne committed:
Remove unnecessary done

RealOrangeOne committed:
Add more general test

RealOrangeOne committed:
More timeout

RealOrangeOne committed:
Refactor mocks to not require mockery internally

RealOrangeOne committed:
Fix mock component

RealOrangeOne committed:
Update ListViewDataSource mock

RealOrangeOne committed:
update requireNativeComponent

RealOrangeOne committed:
add RefreshControl mock

RealOrangeOne committed:
cleanup dependencies

RealOrangeOne committed:
fix lint

RealOrangeOne committed:
QUICKLY

RealOrangeOne committed:
Added sinon for spying

RealOrangeOne committed:
Add test cases for native modules

RealOrangeOne committed:
Test mocked components

RealOrangeOne committed:
Better check module requires

RealOrangeOne committed:
Make images requirable

RealOrangeOne committed:
FIx mock component display name

RealOrangeOne committed:
Fix linting

RealOrangeOne committed:
Start rendering actual components

RealOrangeOne committed:
Add / setup jsdom

RealOrangeOne committed:
Fix haste map generator

RealOrangeOne committed:
Create expect spy wrapper

RealOrangeOne committed:
Add some tests of actual components

RealOrangeOne committed:
Allow 0.38+

RealOrangeOne committed:
Rearrange mocks and add listview mock

RealOrangeOne committed:
rearrange mocks again

RealOrangeOne committed:
less hax in test

RealOrangeOne committed:
Use less simple tests for ActivityIndicator and ListView

RealOrangeOne committed:
Cleanup listview mock

RealOrangeOne committed:
Validate react-native version

RealOrangeOne committed:
Added remaining basic tests

RealOrangeOne committed:
Import old tests

RealOrangeOne committed:
Cleanup readme

RealOrangeOne committed:
fix lint

RealOrangeOne committed:
Added polyfill for navigator if not defined

RealOrangeOne committed:
Fix coverage

RealOrangeOne committed:
Add test for platform native module

RealOrangeOne committed:
Fix lint

RealOrangeOne committed:
Use flat icons on readme

RealOrangeOne committed:
Add timer for haste map

RealOrangeOne committed:
Increase timeout because node4 is really slow

RealOrangeOne committed:
fix eslint

RealOrangeOne committed:
Merge remote-tracking branch 'origin' into rewrite

RealOrangeOne committed:
Add script to stop requiring

RealOrangeOne committed:
Fix lint and remove unused dep

RealOrangeOne committed:
Add codecov

RealOrangeOne committed:
Use actual codecov link...

RealOrangeOne committed:
Fix markdown syntax

RealOrangeOne committed:
Use actual project paths instead of self

RealOrangeOne committed:
Lock to versions of react-native

RealOrangeOne committed:
Fix glob path when in project tests

RealOrangeOne committed:
Test there are items in haste map

RealOrangeOne committed:
Merge remote-tracking branch 'origin' into rewrite

RealOrangeOne committed:
Add tests for alert

RealOrangeOne committed:
Add alertIOS tests

RealOrangeOne committed:
Add tests for appregistry

RealOrangeOne committed:
Add more UIManager spies

RealOrangeOne committed:
Fix lint

RealOrangeOne committed:
Test AppState

RealOrangeOne committed:
Add functional mock for AsyncStorage

RealOrangeOne committed:
Fix import

RealOrangeOne committed:
Add BackAndroid tests

RealOrangeOne committed:
Add CameraRoll tests

RealOrangeOne committed:
Add clipboard test

RealOrangeOne committed:
Add DatePickerAndroid tests

RealOrangeOne committed:
Add Dimensions tests

RealOrangeOne committed:
Add easing tests

RealOrangeOne committed:
Try running all the tests

RealOrangeOne committed:
Install all at once, rebuild haste afterwards

RealOrangeOne committed:
Loose version numbers

RealOrangeOne committed:
Update test for new workflow

RealOrangeOne committed:
MOAR TESTS

RealOrangeOne committed:
Enable circleci for standard testing

RealOrangeOne committed:
Add circleci image to readme

RealOrangeOne committed:
Increase timeout to a minute, node 4 is SLOW

RealOrangeOne committed:
Cover additional native modules in tests

RealOrangeOne committed:
Improve spy checks

RealOrangeOne committed:
Add GeoLocation tests

RealOrangeOne committed:
Add ImageEditor tests

RealOrangeOne committed:
Add ImagePickerIOS tests

RealOrangeOne committed:
Cover main import handler

RealOrangeOne committed:
Fix platform native modules

RealOrangeOne committed:
Add Platform API tests

RealOrangeOne committed:
test against react-native@0.43

RealOrangeOne committed:
Make test match native modules

RealOrangeOne committed:
43 isnt released yet

RealOrangeOne committed:
Add Keyboard tests

RealOrangeOne committed:
Add ImageStore tests

RealOrangeOne committed:
Add InteractionManager tests

RealOrangeOne committed:
dont require mocking script inside tests

RealOrangeOne committed:
fix strange promise stuff

RealOrangeOne committed:
Fix scrollview mock

RealOrangeOne committed:
Lint scripts

RealOrangeOne committed:
format main script

RealOrangeOne committed:
Test async storage is mocked

RealOrangeOne committed:
Update linter rule

RealOrangeOne committed:
Add jest test run for reference

RealOrangeOne committed:
Add Linking test

RealOrangeOne committed:
Try testing against 0.43

RealOrangeOne committed:
Remove pinned deps, upgrade enzyme

JPeer264 commented:
@RealOrangeOne based on your list above, are just the tests missing? So basically your rewrite is almost ready right?

RealOrangeOne commented:
@JPeer264 i believe so, however it hasn't been tested in a few months, and there have been many versions of `react-native` in that time. Please feel free to try out this rewrite and let me know how it goes!

JPeer264 commented:
Hm, I tried now to set up the tests, but all failed. So I am quite confused - also they failed here on this Travis.

RealOrangeOne commented:
@JPeer264 Looks like an issue related to both dependencies, and `react-native` changing their api (somehow?). Which is strange, seeing as everything worked fine the last time I tested it

JPeer264 commented:
I guess I got the mistake. [Here](https://travis-ci.org/RealOrangeOne/react-native-mock/jobs/220325004#L197) it seems, that it does install `react-native@0.38.1` which requires `react@15.4.0-rc4`. But the given react version is `react@15.4.0` which does not satisfy the peer dependencies. It might be better to remove the `react` export and just install `react-native`. But to ensure that the right peer dependency is installed (luckily `react-native` got `react` as peer), you can use [install-peerdeps](https://github.com/nathanhleung/install-peerdeps), which automatically installs the right `react` version.

RealOrangeOne commented:
That sounds like quite an issue, and likely one end users will have too. Potentially experimenting with moving all react-* dependencies to peer might solve the issue? As people will need them installed to use the library anyway

JPeer264 commented:
Unluckily I saw that you need more than just react. So installing just the peers won't install e.g. `react-dom`. So I guess having it the way you got it now is fine. But still it would be better to add the `react-native`s patch version to ensure that npm will install the right version. I made a new list and came up with some mistakes in the travis env list (first and third seems to differ):

| react native version | react version | link to package.json
| -------------- | ----- | ------------|
| 0.38.1   | 15.4.0-rc.4  | [link](https://github.com/facebook/react-native/blob/bf74d8f93a111c7e8fa0b97f21bd86f67652f30d/package.json)
| 0.39.2   | 15.4.0-rc.4  | [link](https://github.com/facebook/react-native/blob/92818eb8938813b7fb8762e2ec7f3bc5aa69e9b7/package.json)
| 0.40.0   | 15.4.0-rc.4  | [link](https://github.com/facebook/react-native/blob/a9357ccdbd4c3759e583d17e18bf65d212500a56/package.json)
| 0.41.104 | 15.4.0       | [link](https://github.com/facebook/react-native/blob/43334e3e323f4dfe0f2ce70807d95a168433ae43/package.json)
| 0.42.3   | 15.4.1       | [link](https://github.com/facebook/react-native/blob/113ebb3254831a635678522eb75375e67b3c8bb8/package.json)
| 0.43.5   | 16.0.0-alpha.6 | [link](https://github.com/facebook/react-native/blob/b19128e34b09d1040c7241462d1a52dbedd10811/package.json)
| 0.44.3   | 16.0.0-alpha.6 | [link](https://github.com/facebook/react-native/blob/93d33a3d25052cd9127d3c09b4904e9fa002dc16/package.json)
| 0.45.1   | 16.0.0-alpha.6 | [link](https://github.com/facebook/react-native/blob/a8342a9673ab48aa21f69cebf5d26899f37448e0/package.json)
| 0.46.4   | 16.0.0-alpha.12 | [link](https://github.com/facebook/react-native/blob/1e2a924ba60001c6f0587c7561536f00b2922cbf/package.json)
| 0.47.1   | 16.0.0-alpha.12 | [link](https://github.com/facebook/react-native/blob/7442dc889991952f6f3743114ab1152d12cacfa1/package.json)

Furthermore, `react-addons-test-utils` is [deprecated](https://www.npmjs.com/package/react-addons-test-utils) since `react-dom@15.5.0` to `react-dom/test-utils`. So I think there must be a detection in `./build/react-native.js` wether the react version is higher or lower than `15.5.0` or `16`.

jasonfma commented:
Would be reasonable to support up to RN 42 (before the fiber stuff)? Seems super valuable to get this in and then maybe tackle the fiber changes separately?

Also happy to try to help with this too.

jasonfma commented:
@RealOrangeOne I've pulled the rewrite branch and I was trying to get it running for the latest RN (0.49) and react 16. I think I got it working but there is a test failure around the Android back button. I was able to update all packages except for `sinon` because of a `nise` error when running the tests.

When I went to test my changes against RN 0.38, I ran into some issues trying to write a bunch of defensive code for stuff that's not in 0.38 but is required for RN 0.49, and for enzyme adapters with the other react version.

I think it'll be a lot easier to get this rewrite done to focus on the latest RN and react 16 as a major breaking change. If we really need older version support we can look to add that in progressively.

Thoughts?

RealOrangeOne commented:
@jasonfma that sounds like a great idea. I've been out of the `react-native` (and mobile) scene for a while. Pivoted my work rather heavily. Yes I completely agree breaking backwards support is probably ok, as `react-native` has undergone a lot of changes since I wrote this PR. 

I think targetting `>=0.49` is probably ok. Once support for that is working, support for older versions might be possible, and the version mask can be widened slightly. I think it's perfectly fine to drop support for `<0.3*` easily, even early `<0.4*`

Commit:
Update to RN 0.49
- update enzyme and use enzyme's react 16 adapter
- add RN and react devDeps for now (reflects the peerDeps)
- Add newly required `reactNativeVersion` and `DeviceInfo` properties to the `NativeModules`
  - `reactNativeVersion` reads from the RN source
  - `DeviceInfo` is set to iPhone 6

Commit:
Fix BackAndroid and add BackHandler
- BackAndroid is deprecated by BackHandler
- added test for BackHandler
- Remove obsolete test of `addEventListener().remove()`

Commit:
Remove `Navigator` test as it was removed from RN

Commit:
Fix Dimensions test

Commit:
Fix tests for Android progress bar and NativeModules Dimensions

Commit:
Update for RN 0.50
- backwards compatible changes for RN 0.50
- RN 0.50 adds attributes to components and breaks the buildComponentHTML approach to testing

Commit:
Package updates
- TODO lint fixes

Commit:
Automatic lint fixes

Commit:
Manual lint fixes

Commit:
Fix test broken by chai upgrade
- remove unnecessary assertion because it's asserted implicitly in the next assertion

Commit:
More package updates

jasonfma commented:
@RealOrangeOne opened a PR into your `rewrite` branch #151

Commit:
Update travis to test RN 0.49+

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Remove react-addons-test-utils from travis builds

Commit:
Use RN 0.50+
- render of `ProgressBarAndroid` changed from `ActivityIndicator` in RN 0.49 back to  `AndroidProgressBar` in RN 0.50

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Add tests for Node 8 now that it's LTS

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Add back devDeps for RN and react to fix build

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Fix CI to not rely on devDeps react and react-native

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Drop support for Node 4

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Node support
- Update readme to reference Node 6 support
- Update travis to include Node 8.0.0 tests

Commit:
Set base Node 8 test version to 8.9.1

RealOrangeOne committed:
Merge pull request #151 from jasonfma/rewrite

Rewrite for RN 0.50+

edmundito commented on src/haste.js:
```diff
@@ -0,0 +1,51 @@
+var files = glob.sync(path.join(PROJECT_NODE_MODULES, '**/*.js'));
+
+_.forEach(files, function (file) {
+  var matches = providesRegex.exec(fs.readFileSync(file).toString());
```
I ran into an issue here where if the package name ends with `.js` ([example](https://github.com/whitequark/ipaddr.js)) it will cause the install script to fail. So my workaround was to check if the file is a actual file first. Additionally  I changed the wildcard above to `*/**/*.js`.

nick-rex commented:
Hi there, what is required to get this over the line? Happy to help if it is of any use

Commit:
Merge branch 'master' of github.com:RealOrangeOne/react-native-mock into rewrite

# Conflicts:
#	src/components/View.js
#	src/propTypes/LayoutPropTypes.js
#	src/react-native.js

jasonfma commented:
I think the core APIs that need to be tested are as follows:

- [ ] AccessibilityInfo
- [ ] ActionSheetIOS
- [ ] Animated
- [ ] LayoutAnimation
- [ ] ListViewDataSource
- [ ] NetInfo
- [ ] PanResponder
- [ ] PermissionsAndroid
- [ ] PixelRatio
- [ ] PushNotificationIOS
- [ ] Settings
- [ ] Share
- [ ] StatusBarIOS
- [ ] Systrace
- [ ] TimePickerAndroid
- [ ] ToastAndroid
- [ ] Transforms
- [ ] Vibration
- [ ] VibrationIOS

jasonfma commented:
I believe the way that RN is doing it with jest is a good guideline of how to get these APIs working.
https://github.com/facebook/react-native/blob/master/jest/setup.js

RealOrangeOne commented:
Theoretically, it should be a simple case of stubbing out the native modules related to those APIs, and possibly make them return sane values if required. I'm hoping none of them require a full stub

mnt committed:
Enable nodir for glob.sync

RealOrangeOne committed:
Merge pull request #160 from mnt/rewrite

Enable nodir for glob.sync

nijolas reviewed (commented):
What is the broken test about? "Cannot find module 'react-native/package'" on `npm install` is the error. Anything to do with this issue by any chance? https://github.com/oblador/react-native-vector-icons/issues/626#issuecomment-357405396

Edit: I completely misunderstood that error - please see https://github.com/RealOrangeOne/react-native-mock/pull/161 for a proposed fix from me

nijolas reviewed (changes requested):
Please incorporate #161 to fix Circle CI tests
